File: backend/src/application/places/place.mapper.ts
/* Pure mappers: Prisma rows -> frontend DTOs. Logic mirrors the mock
   toCategoryDto / toCard / toDetail in web/src/shared/api/placesApi.ts so the
   HTTP responses are byte-identical. */

import type { Category, Place, PlaceCategory, PlacePhoto } from '@prisma/client'
import type { CategoryDto, CategoryId, PlaceCardDto, PlaceDetailDto } from '../../contracts/dto/place.dto'

export type PlaceWithRelations = Place & {
  categories: (PlaceCategory & { category: Category })[]
  photos?: PlacePhoto[]
}

export function toCategoryDto(c: Category): CategoryDto {
  return {
    id: c.id as CategoryId,
    label: c.label,
    short: c.short,
    color: c.cssvar,
    glyph: c.id,
  }
}

function primaryCategory(p: PlaceWithRelations): Category {
  const link = p.categories.find((pc) => pc.primary) ?? p.categories[0]
  return link.category
}

export function toCard(p: PlaceWithRelations): PlaceCardDto {
  return {
    id: p.id,
    slug: p.slug,
    name: p.name,
    category: toCategoryDto(primaryCategory(p)),
    neighborhood: p.neighborhood,
    savesCount: p.savesCount,
    isFree: p.isFree,
    tags: p.tags,
    lat: p.lat,
    lng: p.lng,
  }
}

export function toDetail(p: PlaceWithRelations): PlaceDetailDto {
  const q = `${p.lat},${p.lng}`
  const photos = p.photos && p.photos.length
    ? p.photos.map((ph) => ({ url: ph.url }))
    : [{ url: '' }]
  return {
    ...toCard(p),
    note: p.note,
    photos,
    viewsCount: p.viewsCount,
    sharesCount: p.sharesCount,
    contributor: { name: p.contributorName },
    verifiedAt: p.verifiedLabel,
    fieldNotes: { access: p.accessNote, lit: p.litNote, best: p.bestNote },
    appleMapsUrl: `https://maps.apple.com/?ll=${q}&q=${encodeURIComponent(p.name)}`,
    googleMapsUrl: `https://www.google.com/maps/search/?api=1&query=${q}`,
  }
}

export const PLACE_INCLUDE = {
  categories: { include: { category: true } },
  photos: { orderBy: { sort: 'asc' as const } },
}

File: backend/src/main/java/ee/gemspot/api/mapper/PlaceMapper.java
package ee.gemspot.api.mapper;

import ee.gemspot.api.domain.Category;
import ee.gemspot.api.domain.Place;
import ee.gemspot.api.domain.PlaceCategory;
import ee.gemspot.api.dto.CategoryDto;
import ee.gemspot.api.dto.ContributorDto;
import ee.gemspot.api.dto.FieldNotesDto;
import ee.gemspot.api.dto.PhotoDto;
import ee.gemspot.api.dto.PlaceCardDto;
import ee.gemspot.api.dto.PlaceDetailDto;
import org.springframework.stereotype.Component;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Pure mappers: JPA entities -> frontend DTOs. Logic mirrors the Nest
 * toCategoryDto / toCard / toDetail (backend/src/application/places/place.mapper.ts)
 * so the HTTP responses are byte-identical to the mock + Nest backend.
 */
@Component
public class PlaceMapper {

    public CategoryDto toCategoryDto(Category c) {
        return new CategoryDto(
                c.getId(),
                c.getLabel(),
                c.getShortLabel(),
                c.getCssvar(),
                c.getId()
        );
    }

    public Category primaryCategory(Place p) {
        return p.getCategories().stream()
                .filter(PlaceCategory::isPrimary)
                .findFirst()
                .orElse(p.getCategories().get(0))
                .getCategory();
    }

    public PlaceCardDto toCard(Place p) {
        return new PlaceCardDto(
                p.getId(),
                p.getSlug(),
                p.getName(),
                toCategoryDto(primaryCategory(p)),
                p.getNeighborhood(),
                null,                   // distanceKm — omitted via NON_NULL
                p.getSavesCount(),
                p.isFree(),
                null,                   // thumbUrl — omitted via NON_NULL
                p.getTags(),
                null,                   // isSaved — omitted via NON_NULL
                p.getLat(),
                p.getLng()
        );
    }

    public PlaceDetailDto toDetail(Place p) {
        // q mirrors Nest `${p.lat},${p.lng}` (JS number->string == JDK shortest-repr Double.toString).
        String q = doubleToJs(p.getLat()) + "," + doubleToJs(p.getLng());

        List<PhotoDto> photos = p.getPhotos().stream()
                .sorted(Comparator.comparingInt(ph -> ph.getSort()))
                .map(ph -> new PhotoDto(ph.getUrl()))
                .collect(Collectors.toList());
        if (photos.isEmpty()) {
            photos = List.of(new PhotoDto("")); // {url:""} placeholder, matches Nest
        }

        return new PlaceDetailDto(
                p.getId(),
                p.getSlug(),
                p.getName(),
                toCategoryDto(primaryCategory(p)),
                p.getNeighborhood(),
                null,                   // distanceKm
                p.getSavesCount(),
                p.isFree(),
                null,                   // thumbUrl
                p.getTags(),
                null,                   // isSaved
                p.getLat(),
                p.getLng(),
                p.getNote(),
                photos,
                p.getViewsCount(),
                p.getSharesCount(),
                new ContributorDto(p.getContributorName()),
                p.getVerifiedLabel(),
                new FieldNotesDto(p.getAccessNote(), p.getLitNote(), p.getBestNote()),
                "https://maps.apple.com/?ll=" + q + "&q=" + encodeURIComponent(p.getName()),
                // googleMapsUrl query is raw (NOT url-encoded) in Nest — keep raw.
                "https://www.google.com/maps/search/?api=1&query=" + q
        );
    }

    /** Matches JS Number->String shortest round-trip (JDK 19+ Double.toString). */
    private static String doubleToJs(double d) {
        return Double.toString(d);
    }

    /** Matches JS encodeURIComponent: space -> %20 (URLEncoder emits + ; fix it). */
    private static String encodeURIComponent(String s) {
        return URLEncoder.encode(s, StandardCharsets.UTF_8).replace("+", "%20");
    }
}
